tools/slo-generator/slo_generator/report.py

```diff
@@ -0,0 +1,274 @@
+# Copyright 2019 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#            http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+`report.py`
+Report utilities.
+"""
+
+import logging
+from dataclasses import dataclass, asdict, fields
+from slo_generator import utils
+
+LOGGER = logging.getLogger(__name__)
+
+
+@dataclass(init=False)
+class SLOReport:
+    """SLO report dataclass. Compute an SLO report out of an SLO config and an
+    Error Budget Policy step.
+
+    Args:
+        config (dict): SLO configuration.
+        step (dict): Error budget policy step configuration.
+        timestamp (int): Timestamp.
+        client (obj): Existing backend client.
+        delete (bool): Backend delete action.
+    """
+    # pylint: disable=too-many-instance-attributes
+
+    # SLO
+    service_name: str
+    feature_name: str
+    slo_name: str
+    slo_target: float
+    slo_description: str
+    sli_measurement: float = 0
+    bad_events_count: int = 0
+    good_events_count: int = 0
+    gap: float
+
+    # Error budget
+    error_budget_policy_step_name: str
+    error_budget_target: float
+    error_budget_measurement: float
+    error_budget_burn_rate: float
+    error_budget_minutes: float
+    error_budget_remaining_minutes: float
+    error_minutes: float
+
+    # Error Budget step config
+    timestamp: int
+    timestamp_human: str
+    window: int
+    alert: bool
+    alerting_burn_rate_threshold: float
+    consequence_message: str
+
+    def __init__(self, config, step, timestamp, client=None, delete=False):
+
+        # Init dataclass fields from SLO config and Error Budget Policy
+        self.__set_fields(**config,
+                          **step,
+                          lambdas={
+                              'slo_target': float,
+                              'alerting_burn_rate_threshold': int
+                          })
+
+        # Set other fields
+        self.window = int(step['measurement_window_seconds'])
+        self.timestamp = int(timestamp)
+        self.timestamp_human = utils.get_human_time(timestamp)
+
+        # Get backend results
+        result = self.run_backend(config, client=client, delete=delete)
+        if result:
+            self.build(step, result)
+
+    def build(self, step, result):
+        """Compute all data necessary for the SLO report.
+
+        Args:
+            step (dict): Error Budget Policy step configuration.
+            result (obj): Backend result.
+            delete (bool, optional): Delete mode enabled.
```

Comment:
This isn't an arg?

Reply:
Good catch, had moved this arg to the `run_backend` function. Also, updated the docstring for `run_backend` that included a non-existent argument.